include/deal.II/non_matching/coupling.h

```diff
@@ -0,0 +1,348 @@
+
+  /**
+   * Create a coupling mass matrix for non-matching, overlapping grids.
+   *
+   * Given two non-matching triangulations, representing the domains $\Omega$
+   * and $B$, with $B \subseteq \Omega$, and two finite element spaces
+   * $V(\Omega)$ and $Q(B)$, compute the matrix
+   *
+   * \f[
+   * M_{ij} := \int_{B} v_i(x) w_j(x) dx, \quad \foreach v_i \in V(\Omega), w_j \in Q(B)
+   * \f]
+   *
+   * where $V(\Omega)$ is the finite element space associated with `space_dh`
+   * (or part of it, if specified in `space_comps`), while $Q(B)$ is the finite
+   * element space associated with `immersed_dh` (or part of it, if specified
+   * in a `immersed_comps`).
+   *
+   * The corresponding `sparsity` can be computed by calling the
+   * make_coupling_sparsity_pattern function. The elements of the matrix are
+   * computed by locating the position of quadrature points defined on elements
+   * of $B$ with respect to the embedding triangulation $\Omega$.
+   *
+   * The `space_comps` and `immersed_comps` masks are assumed to be ordered in
+   * the same way: the first component of `space_comps` will couple with the
+   * first component of `immersed_comps`, the second with the second, and so
+   * on. If one of the two masks has more non-zero entries w.r.t. the other,
+   * then the excess components will be ignored.
+   *
+   * If the domain $B$ does not fall withing $\Omega$, an exception will be
+   * thrown by the algorithm that computes the quadrature point locations.
+   *
+   * For both spaces, it is possible to specify a custom Mapping, which
+   * defaults to StaticMappingQ1 for both.
+   *
+   * @author Luca Heltai, 2018
+   */
+  template<int dim0, int dim1, int spacedim, typename Matrix>
+  void create_coupling_mass_matrix(const DoFHandler<dim0, spacedim> &space_dh,
+                                   const DoFHandler<dim1, spacedim> &immersed_dh,
+                                   const Quadrature<dim1>           &quad,
+                                   Matrix                           &matrix,
+                                   const ConstraintMatrix           &constraints = ConstraintMatrix(),
+                                   const ComponentMask              &space_comps = ComponentMask(),
+                                   const ComponentMask              &immersed_comps = ComponentMask(),
+                                   const Mapping<dim0, spacedim>    &space_mapping = StaticMappingQ1<dim0,spacedim>::mapping,
+                                   const Mapping<dim1, spacedim>    &immersed_mapping = StaticMappingQ1<dim1, spacedim>::mapping);
+
+
+
+
+// === inline and template functions ===
+
+
+
+  template<int dim0, int dim1, int spacedim, typename Sparsity>
+  void create_coupling_sparsity_pattern(const DoFHandler<dim0, spacedim> &space_dh,
+                                        const DoFHandler<dim1, spacedim> &immersed_dh,
+                                        const Quadrature<dim1>           &quad,
```

Comment:
these are big functions you're putting into header files here. Couldn't they go into a .cc file, with instantiations for a specific set of template arguments?